pkg/app/pipedv1/plugin/kubernetes/deployment/server.go

```diff
@@ -290,17 +295,85 @@ func (a *DeploymentService) executeK8sSyncStage(ctx context.Context, lp logpersi
 		return model.StageStatus_STAGE_FAILURE
 	}
 
+	// Create the kubectl wrapper for the target cluster.
+	kubectl := provider.NewKubectl(kubectlPath)
+
 	// Create the applier for the target cluster.
-	applier := provider.NewApplier(provider.NewKubectl(kubectlPath), cfg.Spec.Input, deployTargetConfig, a.logger)
+	applier := provider.NewApplier(kubectl, cfg.Spec.Input, deployTargetConfig, a.logger)
 
 	// Start applying all manifests to add or update running resources.
 	if err := applyManifests(ctx, applier, manifests, cfg.Spec.Input.Namespace, lp); err != nil {
 		lp.Errorf("Failed while applying manifests (%v)", err)
 		return model.StageStatus_STAGE_FAILURE
 	}
 
-	// TODO: implement prune resources
+	// TODO: treat the stage options specified under "with"
+	if !cfg.Spec.QuickSync.Prune {
+		lp.Info("Resource GC was skipped because sync.prune was not configured")
+		return model.StageStatus_STAGE_SUCCESS
+	}
+
+	// Wait for all applied manifests to be stable.
+	// In theory, we don't need to wait for them to be stable before going to the next step
+	// but waiting for a while reduces the number of Kubernetes changes in a short time.
+	lp.Info("Waiting for the applied manifests to be stable")
+	select {
+	case <-time.After(15 * time.Second):
+		break
+	case <-ctx.Done():
+		break
+	}
+
+	lp.Info("Start finding all running resources but no longer defined in Git")
+
+	namespacedLiveResources, err := kubectl.GetAll(ctx, deployTargetConfig.KubeConfigPath,
+		"",
+		fmt.Sprintf("%s=%s", provider.LabelManagedBy, provider.ManagedByPiped),
+		fmt.Sprintf("%s=%s", provider.LabelApplication, input.GetDeployment().GetApplicationId()),
+	)
+	if err != nil {
+		lp.Errorf("Failed while listing all resources (%v)", err)
+		return model.StageStatus_STAGE_FAILURE
+	}
+
+	clusterLiveResources, err := kubectl.GetAllClusterScoped(ctx, deployTargetConfig.KubeConfigPath,
```

Comment:
[nit] clusterLiveResources -> clusterScopedLiveResources

Reply:
applied on this commit
https://github.com/pipe-cd/pipecd/pull/5455/commits/6392bca42cf4d162ca0ac380d488281d90ae1cdd